### PR TITLE
GCP Cloud Run hub and spoke implementation

### PR DESCRIPTION
**Pull request by chrismaree**
This PR adds a script to execute batch Cloud Run jobs based off a global config file. This hub script pulls down a global config file stored in google cloud storage buckets and then iteratively runs cloud run jobs for each config object. This enables the bot infrastructure to scale by simply adding new bots to this config object. 

Additionally, this script overalls the `CloudResponse.js` script used to respond to cloud run calls in a more traditional async way. Specifically, this PR implements a `execShellCommand` function that returns a promise for a child process. 

Note that the scripts created in this PR are still poorly organized and are currently sitting in the `/reporters/cloud-run-scripts/` directory. For now, this is fine but this must be moved to a more general directory along with fixing issue: #1530 

This PR does not contain the required stateful logic needed to enable this hub and spoke approach to work fully with our current setup. In particular, the monitor bot needs to know the block number it last used when running a query. This will require a connection to a DB and is left for a separate PR in issue #1640

close #1629

### CODE REVIEW

File: reporters/cloud-run-scripts/CloudResponse.js [hunk 1]
```diff
@@ -1,54 +1,68 @@
...
+    }
+    console.log("Process exited with no error");
+
+    res.status(200).send({ message: "Process exited correctly!", error: null });
+  } catch (error) {
+    res.status(400).send({ message: "Process exited with error!", error: error });
+  }
 });
 
+function execShellCommand(cmd, inputEnv) {
```
Comment (chrismaree) [inline reporters/cloud-run-scripts/CloudResponse.js:59]: The child process logic was refined to run as a promise. Previously this was not strictly blocking logic and there was inconsistent behavior with executing multiple in parallel. This new implementation returns a promise that will only resolve when the process exits.

File: package.json [hunk 2]
```diff
@@ -60,6 +60,7 @@
     "express": "^4.17.1",
     "ganache-cli": "^6.7.0",
     "gmail-send": "^1.2.14",
+    "google-auth-library": "^6.0.2",
```
Comment (chrismaree) [inline package.json:63]: Google library for getting access headers to enable cloud run instances to call _other_ cloud run instances. This lets the "hub" call the "spokes".

**Comment by chrismaree** [comment]
> This is all super cool!
> 
> A few architectural things I wanted to ask/mention:
> 
> 1. In this setup, how do we give bots some sort of persistent state/cache? Would this happen in the spoke or the hub?
> 2. As a user/debugger, it might be useful to be able to easily run a single spoke. I think that this would require creating a new config, then sending a custom request pointing to that config. That's not too bad, but if the config lookup were done in the spokes themselves, one could send a simple request with the config name/bucket and bot name to get it to run with no requirement to mess with configs. I'm not suggesting that we do it that way -- in fact, I think we shouldn't. It would definitely be premature optimization. But on that note, just wanted to hear your thoughts on why you decided to do the lookup in the hub rather than the spokes? One obvious advantage is that the spokes remain pretty generalized and all of the gcloud specifics are located in the hub.

Thanks for the thoughtful feedback @mrice32 regarding your questions:
1. The persistent state will come in by connecting the hub to a NoSQL store. The hub will then start off by querying the last block number it ran from the db, and appending this to the configs sent to all the bots to inform them of where to start running from. At the end of the execution, the hub will update the NoSQL store with the current block number (or more precisely the block number queried at the start of the execution so no blocks are missed).
2. A single spoke can be run in one of two ways: a) calling the spoke directly debugger -> spoke (basically mocking the hub) or b) as you suggested, creating a custom config that goes debugger -> hub -> spoke. Both of these can be tested in unit tests and we can set up integration testing between the layers. I do agree though that it is a little bit clunky for a debugger.

The reason I set it up like this keep the spokes completely stateless and generalized. They don't need to know anything other than "run this command with these environment variables". This is the same reason that (1) has the Hub storing the block numbers, rather than each spoke. This abstracts away the logic from how we call the spokes are called from the implementation of the spokes themselves, leaving the spokes to be truly stateless and serverless. They simply do one task: run a given piece of code and exit. In future, we could choose to replace the hub with any kind of other logic (cloud scheduler with a Json object, for example) and the spoke won't need to change at all.

Please let me know what you think :)

File: reporters/cloud-run-scripts/CloudRunnerHub.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+// GCP helpers
+const { GoogleAuth } = require("google-auth-library");
+const auth = new GoogleAuth();
+const { Storage } = require("@google-cloud/storage");
+const storage = new Storage();
+
+// Helpers
+// TODO integrate with winston logger.
+// const { Logger, waitForLogger } = require("../../financial-templates-lib/logger/Logger");
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply reporters/cloud-run-scripts/CloudRunnerHub.js:38]: ye, I'm not 100% sure just yet to be honest. the main reason I wanted to put Winston in here was to catch errors and log them in our error handling pipeline upstream. Basically, the hub could tell us if something `errors` out. For example, what happens if the hub cant get configs? it should send a Winston `error`.

**Comment by mrice32** [comment]
This all sounds good to me. I think this implementation makes sense. And it's totally fine for it to be a _bit_ clunky to run an individual spoke for debugging.

> the hub will update the NoSQL store with the current block number (or more precisely the block number queried at the start of the execution so no blocks are missed).

One minor question: how is the hub going to know the block number that the spoke searched up _to_? For instance, let's say the NoSQL db says we should start at block 10. That information is propagated to the spoke and it searches from block 10 to `latest`. How will the hub know what block the spoke saw as "latest"? If it doesn't know, isn't it inevitable that it will either risk double querying blocks or skipping blocks on the next execution?

**Comment by chrismaree** [comment]
> @chrismaree This all sounds good to me. I think this implementation makes sense. And it's totally fine for it to be a _bit_ clunky to run an individual spoke for debugging.
> 
> > the hub will update the NoSQL store with the current block number (or more precisely the block number queried at the start of the execution so no blocks are missed).
> 
> One minor question: how is the hub going to know the block number that the spoke searched up _to_? For instance, let's say the NoSQL db says we should start at block 10. That information is propagated to the spoke and it searches from block 10 to `latest`. How will the hub know what block the spoke saw as "latest"? If it doesn't know, isn't it inevitable that it will either risk double querying blocks or skipping blocks on the next execution?

Ye, this is a good point. I think that this is only possible if we _only_ provide a `STARTING_BLOCK_NUMBER`. A solution is to provide both a starting and ending block numbers. My current thinking of this flow is as follows:
1) update the disputer bot to take in two parameters: `STARTING_BLOCK_NUMBER` and `ENDING_BLOCK_NUMBER`. 
2) at the start of the hub's loop it queries the current block number. Call this `latestBlockNumber`
3) the hub fetches the last block number seen from the NoSQL store. Call this `lastBlockNumber`
3) append the `STARTING_BLOCK_NUMBER=lastBlockNumber` to configs as the last block number from the SQL store and set the `ENDING_BLOCK_NUMBERR=latestBlockNumber` as the last block the hub has seen.
4) execute the spoke logic for all the hubs
5) set the NoSQL store last seen block number to `latestBlockNumber` (same value as queried at the start of the loop and used as the `ENDING_BLOCK_NUMBER` in the spokes.

This flow ensures that the spokes run over a given time frame, irrespective of their local impression of block numbers. As a result the hub is responsible for keeping everything in line and only the hub needs to know these block numbers. By the hub specifying both a start and end block the case you described is mitigated.

**Pull request merged**